query-engine/connector-test-kit-rs/query-test-macros/src/ensure_db_names.rs

```diff
@@ -0,0 +1,57 @@
+use std::collections::HashSet;
+use std::sync::Mutex;
+
+// Ensures unique test database names.
+// Reusing the same name more than once causes random test failures, retries and overall flakiness.
+pub struct UniqueTestDatabaseNames {
+    names: Mutex<HashSet<String>>,
+}
+
+impl UniqueTestDatabaseNames {
+    pub fn new() -> Self {
+        Self {
+            names: Mutex::new(HashSet::new()),
+        }
+    }
+
+    pub fn ensure_unique(
+        self: &UniqueTestDatabaseNames,
+        test_database_name: &String,
+        suite_name: &String,
+        test_name: &String,
+    ) {
+        match self.names.lock() {
+            Ok(mut names) => {
+                if names.contains(test_database_name) {
+                    panic!(
+                        "Test database (or schema) names must be unique.\n\
+                         It is concatenation of the test suite and the test case names.\n\
+                         To resolve this error rename them until they are unique.\n\
+                         - Test suite: `{}`\n\
+                         - Test case: `{}`\n\
+                         - Database (or schema): `{}`\n",
+                        suite_name, test_name, test_database_name
+                    );
+                }
+                names.insert(test_database_name.to_string());
+            }
+
+            Err(_) => {
+                // Ignore poisoned RwLock, when another thread has already panicked.
```

Comment:
```suggestion
                // Ignore poisoned Mutex, when another thread has already panicked.
```

Reply:
Yeah, I had a `RwLock` before, but it does not support upgrading from read to write, so had to fall back to using a `Mutex`.

Reply:
I think `Mutex` is fine here. We almost always have to write (in fact, we can only avoid writing in the error case), so using an `RwLock` would've only added extra overhead with no benefit.